<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/08_model_refinement.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB08 — Modelagem Supervisionada e Validação Temporal

## 1. Contexto

Este notebook realiza a etapa principal de modelagem supervisionada do pipeline.

A entrada desta etapa é a base supervisionada de antecipação construída no NB06, correspondente ao cenário revisado `primary_K24_H12`.

Nesse cenário:

- \(K = 24\) janelas define a região pré-crítica `BEFORE`;
- \(H = 12\) janelas define o horizonte supervisionado de antecipação;
- o horizonte preditivo corresponde a 60 minutos;
- a janela pré-crítica corresponde a 120 minutos;
- a classe positiva indica que uma janela `DURING` ocorrerá dentro das próximas 12 janelas.

Essa formulação substitui a versão anterior `K12_H12`, na qual a classe positiva coincidia diretamente com o estado `BEFORE`.

## 2. Objetivo

O objetivo deste notebook é:

- carregar a base supervisionada revisada;
- selecionar atributos explicativos sem vazamento de informação;
- treinar modelos supervisionados de classificação binária;
- tratar o desbalanceamento de classes;
- usar validação temporal por partições progressivas como avaliação principal;
- manter o corte temporal fixo apenas como análise complementar;
- reportar médias e desvios-padrão das métricas por fold;
- avaliar sensibilidade a diferentes cortes temporais;
- gerar artefatos rastreáveis para interpretação e consolidação final.

## 3. Papel no Pipeline Atualizado

O NB08 é a etapa responsável por verificar se a formulação supervisionada construída no NB06 produz sinal preditivo defensável.

Diferentemente de versões anteriores, esta execução considera a validação temporal por \textit{TimeSeriesSplit} como evidência principal de desempenho, pois ela é mais compatível com o uso operacional em séries temporais.

O corte temporal fixo 80/20 é mantido como análise secundária, útil para gerar diagnósticos, figuras e probabilidades previstas, mas não deve ser interpretado isoladamente como estimativa principal de generalização operacional.

## 4. Modelos Avaliados

São avaliados modelos adequados para dados tabulares:

- Regressão Logística;
- Random Forest;
- ExtraTrees;
- HistGradientBoosting.

A Regressão Logística usa padronização por `StandardScaler`, pois é sensível à escala das variáveis.

Os modelos baseados em árvores são treinados sem padronização, pois não dependem da escala das variáveis.

O desbalanceamento é tratado por:

- `class_weight='balanced'` na Regressão Logística;
- `class_weight='balanced_subsample'` em Random Forest e ExtraTrees;
- `sample_weight` no ajuste do HistGradientBoosting.

## 5. Validação Temporal Principal

A avaliação principal é realizada por validação temporal por partições progressivas (`TimeSeriesSplit`).

Para cada modelo, são reportados:

- acurácia média e desvio-padrão;
- precisão média e desvio-padrão;
- recall médio e desvio-padrão;
- F1-score médio e desvio-padrão;
- ROC-AUC médio e desvio-padrão;
- Average Precision médio e desvio-padrão;
- Brier Score médio e desvio-padrão.

A inclusão dos desvios-padrão permite avaliar a estabilidade temporal do desempenho entre folds.

## 6. Corte Temporal Fixo

O corte temporal fixo 80/20 é mantido como análise complementar.

Ele permite:

- comparar com execuções anteriores;
- gerar matriz de confusão;
- gerar curvas ROC e Precision-Recall;
- produzir a série temporal de probabilidades previstas usada no NB09.

Porém, devido à mudança de distribuição entre treino e teste, esse corte deve ser interpretado com cautela.

## 7. Análise de Sensibilidade dos Cortes

Além do corte 80/20, são avaliados cortes cronológicos adicionais:

- 60/40;
- 70/30;
- 80/20.

Essa análise verifica se o desempenho é sensível à posição do ponto de separação entre treino e teste.

## 8. Importância dos Atributos

A importância por permutação é calculada para o modelo selecionado.

Nesta versão, a análise é tratada como interpretativa, e não como critério de seleção de atributos.

Sempre que possível, também é estimada a importância por permutação nos folds da validação temporal, reduzindo a dependência de um único corte fixo.

## 9. Saídas Esperadas

Ao final deste notebook, espera-se obter:

- `08_supervised_modeling_summary.json`;
- `08_tscv_metrics_by_fold.csv`;
- `08_tscv_summary.csv`;
- `08_fixed_split_metrics.csv`;
- `08_split_sensitivity_metrics.csv`;
- `08_permutation_importance_fixed.csv`;
- `08_permutation_importance_tscv.csv`;
- `08_selected_model_test_predictions.parquet`;
- `08_selected_model.joblib`;
- `08_selected_feature_columns.json`.

## 10. Observações Metodológicas

Esta etapa deve evitar qualquer interpretação forçada dos resultados.

Assim:

- a validação temporal é tratada como resultado principal;
- o corte fixo é tratado como análise complementar;
- médias são acompanhadas de desvios-padrão;
- o modelo selecionado é escolhido com base no desempenho médio em validação temporal;
- diferenças de pré-processamento entre modelos são explicitadas;
- probabilidades previstas são interpretadas como escores de risco, cuja calibração formal será avaliada no NB09.

In [1]:
# ============================================================
# NB08 — Modelagem Supervisionada e Validação Temporal
# Versão revisada após NB06 primary_K24_H12
# - TSCV como avaliação principal
# - corte fixo como análise complementar
# - média e desvio-padrão por fold
# - HGB com sample_weight
# - importância por permutação com ressalvas
# ============================================================

from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import json
import warnings

import numpy as np
import pandas as pd

from IPython.display import display

from sklearn.base import clone
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import TimeSeriesSplit
from sklearn.utils.class_weight import compute_sample_weight

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, HistGradientBoostingClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    average_precision_score,
    brier_score_loss,
    confusion_matrix,
)

from sklearn.inspection import permutation_importance

import joblib

warnings.filterwarnings("ignore")

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

# -----------------------------
# 0) BOOTSTRAP
# -----------------------------

if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull)...")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso: não foi possível atualizar via git pull:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, ensure_dirs

ensure_dirs()

MODELS_PATH = Path("/content/drive/MyDrive/Mestrado/03-models")
MODELS_PATH.mkdir(parents=True, exist_ok=True)

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH =", REPORTS_PATH)
print("MODELS_PATH =", MODELS_PATH)

def log(msg: str) -> None:
    print(f"[NB08_supervised_modeling] {msg}")


# -----------------------------
# 1) PARÂMETROS E ARQUIVOS
# -----------------------------

WINDOW_MINUTES = 5
N_SPLITS_TSCV = 5
FIXED_SPLIT_FRACTION = 0.80
SPLIT_SENSITIVITY_FRACTIONS = [0.60, 0.70, 0.80]

TARGET_COL = "transition_target"
PROBA_THRESHOLD = 0.50

INPUT_SUPERVISED_FILE = FEATURES_PATH / "anticipation_supervised_dataset.parquet"
NB06_SUMMARY_FILE = REPORTS_PATH / "06_labeling_states_summary.json"

SUMMARY_FILE = REPORTS_PATH / "08_supervised_modeling_summary.json"

TSCV_FOLD_METRICS_FILE = REPORTS_PATH / "08_tscv_metrics_by_fold.csv"
TSCV_SUMMARY_FILE = REPORTS_PATH / "08_tscv_summary.csv"
FIXED_METRICS_FILE = REPORTS_PATH / "08_fixed_split_metrics.csv"
SPLIT_SENSITIVITY_FILE = REPORTS_PATH / "08_split_sensitivity_metrics.csv"

PERM_FIXED_FILE = REPORTS_PATH / "08_permutation_importance_fixed.csv"
PERM_TSCV_FILE = REPORTS_PATH / "08_permutation_importance_tscv.csv"

SELECTED_MODEL_FILE = MODELS_PATH / "08_selected_model.joblib"
SELECTED_FEATURES_FILE = REPORTS_PATH / "08_selected_feature_columns.json"
SELECTED_TEST_PREDICTIONS_FILE = REPORTS_PATH / "08_selected_model_test_predictions.parquet"


# -----------------------------
# 2) LEITURA DA BASE SUPERVISIONADA
# -----------------------------

assert INPUT_SUPERVISED_FILE.exists(), f"Arquivo não encontrado: {INPUT_SUPERVISED_FILE}"

df = pd.read_parquet(INPUT_SUPERVISED_FILE)

assert TARGET_COL in df.columns, f"Coluna alvo ausente: {TARGET_COL}"
assert "bucket_id" in df.columns, "Coluna bucket_id ausente."

df = df.sort_values("bucket_id").reset_index(drop=True)

log(f"Base supervisionada carregada: shape={df.shape}")
log(f"Arquivo: {INPUT_SUPERVISED_FILE}")

if NB06_SUMMARY_FILE.exists():
    nb06_summary = json.loads(NB06_SUMMARY_FILE.read_text(encoding="utf-8"))
    log(f"Resumo do NB06 carregado: {NB06_SUMMARY_FILE}")
else:
    nb06_summary = {}

target_counts = df[TARGET_COL].value_counts().sort_index()
target_distribution = {str(k): int(v) for k, v in target_counts.to_dict().items()}
positive_rate = float(df[TARGET_COL].mean())

print("\n=== Distribuição do alvo ===")
display(target_counts.to_frame("count"))
print(f"Taxa positiva: {positive_rate:.4f}")

if "state" in df.columns:
    print("\n=== Tabela state × target ===")
    display(pd.crosstab(df["state"], df[TARGET_COL]))


# -----------------------------
# 3) SELEÇÃO DE ATRIBUTOS SEM VAZAMENTO
# -----------------------------

EXCLUDE_EXACT = {
    # alvo e derivados do alvo
    "transition_target",
    "transition_label",
    "target_definition",
    "scenario_name",

    # estados e transições
    "state",
    "state_rank",
    "state_window_k",
    "next_bucket_id",
    "next_state",
    "is_last_window",
    "has_valid_next",
    "transition",

    # campos futuros usados apenas para rotulagem retrospectiva
    "future_horizon_valid",
    "future_has_during",
    "future_all_normal",
    "future_path_states",
    "future_first_during_offset",
    "target_horizon_windows",
    "prediction_horizon_minutes",

    # identificadores/tempo absoluto
    "bucket_id",
    "bucket_start_us",
    "bucket_end_us",
    "window_start_us",
    "window_end_us",

    # rótulo crítico instantâneo derivado da segmentação
    "is_critical",
}

EXCLUDE_PREFIXES = (
    "future_",
    "target_",
    "next_",
)

candidate_cols = []

for col in df.columns:
    if col in EXCLUDE_EXACT:
        continue

    if any(col.startswith(prefix) for prefix in EXCLUDE_PREFIXES):
        continue

    if pd.api.types.is_numeric_dtype(df[col]):
        candidate_cols.append(col)

feature_cols = candidate_cols

assert len(feature_cols) > 0, "Nenhum atributo explicativo selecionado."

X = df[feature_cols].copy()
y = df[TARGET_COL].astype(int).copy()

print("\n=== Atributos selecionados ===")
print(f"Total de atributos: {len(feature_cols)}")
for c in feature_cols:
    print("-", c)

feature_selection_info = {
    "total_columns_input": int(df.shape[1]),
    "selected_feature_count": int(len(feature_cols)),
    "selected_features": feature_cols,
    "excluded_exact": sorted(list(EXCLUDE_EXACT)),
    "exclude_prefixes": list(EXCLUDE_PREFIXES),
    "selection_rule": (
        "Numeric columns excluding identifiers, state labels, future horizon fields, "
        "target fields, transition fields, and instantaneous critical label."
    ),
}

SELECTED_FEATURES_FILE.write_text(
    json.dumps(feature_selection_info, indent=2, ensure_ascii=False),
    encoding="utf-8",
)

log(f"Lista de atributos salva: {SELECTED_FEATURES_FILE}")


# -----------------------------
# 4) DEFINIÇÃO DOS MODELOS
# -----------------------------

def build_model_specs():
    specs = {
        "LogisticRegression": {
            "estimator": Pipeline(steps=[
                ("imputer", SimpleImputer(strategy="median")),
                ("scaler", StandardScaler()),
                ("classifier", LogisticRegression(
                    max_iter=5000,
                    class_weight="balanced",
                    solver="lbfgs",
                )),
            ]),
            "use_sample_weight": False,
            "notes": "StandardScaler + class_weight='balanced'",
        },

        "RandomForest": {
            "estimator": Pipeline(steps=[
                ("imputer", SimpleImputer(strategy="median")),
                ("classifier", RandomForestClassifier(
                    n_estimators=300,
                    random_state=SEED,
                    n_jobs=-1,
                    class_weight="balanced_subsample",
                    min_samples_leaf=2,
                )),
            ]),
            "use_sample_weight": False,
            "notes": "class_weight='balanced_subsample'",
        },

        "ExtraTrees": {
            "estimator": Pipeline(steps=[
                ("imputer", SimpleImputer(strategy="median")),
                ("classifier", ExtraTreesClassifier(
                    n_estimators=300,
                    random_state=SEED,
                    n_jobs=-1,
                    class_weight="balanced",
                    min_samples_leaf=2,
                )),
            ]),
            "use_sample_weight": False,
            "notes": "class_weight='balanced'",
        },

        "HistGradientBoosting": {
            "estimator": Pipeline(steps=[
                ("imputer", SimpleImputer(strategy="median")),
                ("classifier", HistGradientBoostingClassifier(
                    max_iter=300,
                    learning_rate=0.05,
                    max_leaf_nodes=31,
                    min_samples_leaf=20,
                    random_state=SEED,
                )),
            ]),
            "use_sample_weight": True,
            "notes": "sample_weight=compute_sample_weight('balanced', y_train)",
        },
    }

    return specs


MODEL_SPECS = build_model_specs()

print("\n=== Modelos avaliados ===")
for name, spec in MODEL_SPECS.items():
    print(f"{name}: {spec['notes']}")


# -----------------------------
# 5) FUNÇÕES DE TREINO E AVALIAÇÃO
# -----------------------------

def fit_estimator(estimator, X_train, y_train, use_sample_weight: bool):
    est = clone(estimator)

    if use_sample_weight:
        weights = compute_sample_weight(class_weight="balanced", y=y_train)
        est.fit(X_train, y_train, classifier__sample_weight=weights)
    else:
        est.fit(X_train, y_train)

    return est


def get_positive_proba(estimator, X_eval):
    if hasattr(estimator, "predict_proba"):
        proba = estimator.predict_proba(X_eval)[:, 1]
        return proba

    if hasattr(estimator, "decision_function"):
        scores = estimator.decision_function(X_eval)
        proba = 1 / (1 + np.exp(-scores))
        return proba

    raise RuntimeError("Estimador não fornece predict_proba nem decision_function.")


def safe_roc_auc(y_true, y_proba):
    if len(np.unique(y_true)) < 2:
        return np.nan
    return float(roc_auc_score(y_true, y_proba))


def safe_average_precision(y_true, y_proba):
    if len(np.unique(y_true)) < 2:
        return np.nan
    return float(average_precision_score(y_true, y_proba))


def compute_metrics(y_true, y_proba, threshold=PROBA_THRESHOLD):
    y_true = np.asarray(y_true).astype(int)
    y_proba = np.asarray(y_proba).astype(float)
    y_pred = (y_proba >= threshold).astype(int)

    tn, fp, fn, tp = confusion_matrix(y_true, y_pred, labels=[0, 1]).ravel()

    return {
        "accuracy": float(accuracy_score(y_true, y_pred)),
        "precision": float(precision_score(y_true, y_pred, zero_division=0)),
        "recall": float(recall_score(y_true, y_pred, zero_division=0)),
        "f1": float(f1_score(y_true, y_pred, zero_division=0)),
        "roc_auc": safe_roc_auc(y_true, y_proba),
        "average_precision": safe_average_precision(y_true, y_proba),
        "brier_score": float(brier_score_loss(y_true, y_proba)),
        "tn": int(tn),
        "fp": int(fp),
        "fn": int(fn),
        "tp": int(tp),
        "support": int(len(y_true)),
        "positives": int(y_true.sum()),
        "positive_rate": float(y_true.mean()) if len(y_true) > 0 else np.nan,
    }


def summarize_metrics(metrics_df, group_col="model"):
    metric_cols = [
        "accuracy",
        "precision",
        "recall",
        "f1",
        "roc_auc",
        "average_precision",
        "brier_score",
    ]

    rows = []

    for name, group in metrics_df.groupby(group_col):
        row = {group_col: name}

        for metric in metric_cols:
            row[f"{metric}_mean"] = float(group[metric].mean())
            row[f"{metric}_std"] = float(group[metric].std(ddof=1)) if len(group) > 1 else 0.0

        row["folds"] = int(len(group))
        row["test_positive_rate_mean"] = float(group["test_positive_rate"].mean())
        row["test_positive_rate_std"] = float(group["test_positive_rate"].std(ddof=1)) if len(group) > 1 else 0.0

        rows.append(row)

    return pd.DataFrame(rows)


# -----------------------------
# 6) VALIDAÇÃO TEMPORAL PRINCIPAL — TimeSeriesSplit
# -----------------------------

tscv = TimeSeriesSplit(n_splits=N_SPLITS_TSCV)

tscv_rows = []

for model_name, spec in MODEL_SPECS.items():
    log(f"TSCV — avaliando {model_name}")

    for fold_id, (train_idx, test_idx) in enumerate(tscv.split(X), start=1):
        X_train = X.iloc[train_idx]
        y_train = y.iloc[train_idx]

        X_test = X.iloc[test_idx]
        y_test = y.iloc[test_idx]

        est = fit_estimator(
            estimator=spec["estimator"],
            X_train=X_train,
            y_train=y_train,
            use_sample_weight=spec["use_sample_weight"],
        )

        y_proba = get_positive_proba(est, X_test)

        metrics = compute_metrics(
            y_true=y_test,
            y_proba=y_proba,
            threshold=PROBA_THRESHOLD,
        )

        row = {
            "validation": "TimeSeriesSplit",
            "model": model_name,
            "fold": int(fold_id),
            "train_start_pos": int(train_idx[0]),
            "train_end_pos": int(train_idx[-1]),
            "test_start_pos": int(test_idx[0]),
            "test_end_pos": int(test_idx[-1]),
            "train_rows": int(len(train_idx)),
            "test_rows": int(len(test_idx)),
            "train_positives": int(y_train.sum()),
            "test_positives": int(y_test.sum()),
            "train_positive_rate": float(y_train.mean()),
            "test_positive_rate": float(y_test.mean()),
        }

        row.update(metrics)
        tscv_rows.append(row)

tscv_metrics_df = pd.DataFrame(tscv_rows)
tscv_summary_df = summarize_metrics(tscv_metrics_df, group_col="model")

# Critério de ordenação: F1 médio, depois ROC-AUC médio, depois AP média.
tscv_summary_df = tscv_summary_df.sort_values(
    ["f1_mean", "roc_auc_mean", "average_precision_mean"],
    ascending=[False, False, False],
).reset_index(drop=True)

print("\n=== Métricas por fold — TSCV ===")
display(tscv_metrics_df)

print("\n=== Resumo TSCV — média e desvio-padrão ===")
display(tscv_summary_df)

tscv_metrics_df.to_csv(TSCV_FOLD_METRICS_FILE, index=False)
tscv_summary_df.to_csv(TSCV_SUMMARY_FILE, index=False)

log(f"Métricas TSCV por fold salvas: {TSCV_FOLD_METRICS_FILE}")
log(f"Resumo TSCV salvo: {TSCV_SUMMARY_FILE}")


# -----------------------------
# 7) SELEÇÃO DO MODELO PRINCIPAL
# -----------------------------

selected_model_name = str(tscv_summary_df.iloc[0]["model"])
selected_spec = MODEL_SPECS[selected_model_name]

log(f"Modelo selecionado pelo critério TSCV: {selected_model_name}")

selected_tscv_summary = tscv_summary_df.iloc[0].to_dict()


# -----------------------------
# 8) CORTE TEMPORAL FIXO 80/20 — ANÁLISE COMPLEMENTAR
# -----------------------------

split_idx = int(np.floor(len(df) * FIXED_SPLIT_FRACTION))

X_train_fixed = X.iloc[:split_idx]
y_train_fixed = y.iloc[:split_idx]

X_test_fixed = X.iloc[split_idx:]
y_test_fixed = y.iloc[split_idx:]

fixed_rows = []
fixed_estimators = {}

for model_name, spec in MODEL_SPECS.items():
    log(f"Corte fixo 80/20 — avaliando {model_name}")

    est = fit_estimator(
        estimator=spec["estimator"],
        X_train=X_train_fixed,
        y_train=y_train_fixed,
        use_sample_weight=spec["use_sample_weight"],
    )

    y_proba = get_positive_proba(est, X_test_fixed)

    metrics = compute_metrics(
        y_true=y_test_fixed,
        y_proba=y_proba,
        threshold=PROBA_THRESHOLD,
    )

    row = {
        "validation": "fixed_temporal_split",
        "split_fraction": float(FIXED_SPLIT_FRACTION),
        "model": model_name,
        "train_rows": int(len(X_train_fixed)),
        "test_rows": int(len(X_test_fixed)),
        "train_positives": int(y_train_fixed.sum()),
        "test_positives": int(y_test_fixed.sum()),
        "train_positive_rate": float(y_train_fixed.mean()),
        "test_positive_rate": float(y_test_fixed.mean()),
    }

    row.update(metrics)
    fixed_rows.append(row)
    fixed_estimators[model_name] = est

fixed_metrics_df = pd.DataFrame(fixed_rows).sort_values(
    ["f1", "roc_auc", "average_precision"],
    ascending=[False, False, False],
).reset_index(drop=True)

print("\n=== Corte temporal fixo 80/20 ===")
display(fixed_metrics_df)

fixed_metrics_df.to_csv(FIXED_METRICS_FILE, index=False)
log(f"Métricas do corte fixo salvas: {FIXED_METRICS_FILE}")


# -----------------------------
# 9) ANÁLISE DE SENSIBILIDADE DOS CORTES TEMPORAIS
# -----------------------------

split_sensitivity_rows = []

for split_fraction in SPLIT_SENSITIVITY_FRACTIONS:
    cut_idx = int(np.floor(len(df) * split_fraction))

    X_train_cut = X.iloc[:cut_idx]
    y_train_cut = y.iloc[:cut_idx]

    X_test_cut = X.iloc[cut_idx:]
    y_test_cut = y.iloc[cut_idx:]

    for model_name, spec in MODEL_SPECS.items():
        log(f"Sensibilidade corte {split_fraction:.2f} — {model_name}")

        est = fit_estimator(
            estimator=spec["estimator"],
            X_train=X_train_cut,
            y_train=y_train_cut,
            use_sample_weight=spec["use_sample_weight"],
        )

        y_proba = get_positive_proba(est, X_test_cut)

        metrics = compute_metrics(
            y_true=y_test_cut,
            y_proba=y_proba,
            threshold=PROBA_THRESHOLD,
        )

        row = {
            "validation": "split_sensitivity",
            "split_fraction": float(split_fraction),
            "model": model_name,
            "train_rows": int(len(X_train_cut)),
            "test_rows": int(len(X_test_cut)),
            "train_positives": int(y_train_cut.sum()),
            "test_positives": int(y_test_cut.sum()),
            "train_positive_rate": float(y_train_cut.mean()),
            "test_positive_rate": float(y_test_cut.mean()),
        }

        row.update(metrics)
        split_sensitivity_rows.append(row)

split_sensitivity_df = pd.DataFrame(split_sensitivity_rows)

print("\n=== Sensibilidade por corte temporal ===")
display(split_sensitivity_df)

split_sensitivity_df.to_csv(SPLIT_SENSITIVITY_FILE, index=False)
log(f"Sensibilidade de cortes salva: {SPLIT_SENSITIVITY_FILE}")


# -----------------------------
# 10) MODELO SELECIONADO — TREINO FIXO E PREDIÇÕES
# -----------------------------

selected_estimator_fixed = fixed_estimators[selected_model_name]

selected_y_proba_fixed = get_positive_proba(selected_estimator_fixed, X_test_fixed)
selected_y_pred_fixed = (selected_y_proba_fixed >= PROBA_THRESHOLD).astype(int)

pred_cols = [
    "bucket_id",
    "bucket_start_us",
    "state",
    "transition_target",
]

pred_cols_existing = [c for c in pred_cols if c in df.columns]

selected_predictions_df = df.iloc[split_idx:][pred_cols_existing].copy()
selected_predictions_df["y_true"] = y_test_fixed.values
selected_predictions_df["y_proba"] = selected_y_proba_fixed
selected_predictions_df["y_pred"] = selected_y_pred_fixed
selected_predictions_df["model"] = selected_model_name
selected_predictions_df["split_fraction"] = FIXED_SPLIT_FRACTION

selected_predictions_df.to_parquet(
    SELECTED_TEST_PREDICTIONS_FILE,
    compression="snappy",
    index=False,
)

joblib.dump(selected_estimator_fixed, SELECTED_MODEL_FILE)

log(f"Predições do modelo selecionado salvas: {SELECTED_TEST_PREDICTIONS_FILE}")
log(f"Modelo selecionado salvo: {SELECTED_MODEL_FILE}")


# -----------------------------
# 11) IMPORTÂNCIA POR PERMUTAÇÃO — CORTE FIXO
# -----------------------------

log(f"Calculando importância por permutação no corte fixo para {selected_model_name}")

perm_fixed = permutation_importance(
    selected_estimator_fixed,
    X_test_fixed,
    y_test_fixed,
    scoring="average_precision",
    n_repeats=10,
    random_state=SEED,
    n_jobs=-1,
)

perm_fixed_df = pd.DataFrame({
    "feature": feature_cols,
    "importance_mean": perm_fixed.importances_mean,
    "importance_std": perm_fixed.importances_std,
    "source": "fixed_temporal_split",
    "model": selected_model_name,
    "scoring": "average_precision",
}).sort_values("importance_mean", ascending=False).reset_index(drop=True)

print("\n=== Importância por permutação — corte fixo ===")
display(perm_fixed_df.head(20))

perm_fixed_df.to_csv(PERM_FIXED_FILE, index=False)
log(f"Importância por permutação no corte fixo salva: {PERM_FIXED_FILE}")


# -----------------------------
# 12) IMPORTÂNCIA POR PERMUTAÇÃO — TSCV DO MODELO SELECIONADO
# -----------------------------

perm_tscv_rows = []

log(f"Calculando importância por permutação nos folds TSCV para {selected_model_name}")

for fold_id, (train_idx, test_idx) in enumerate(tscv.split(X), start=1):
    X_train = X.iloc[train_idx]
    y_train = y.iloc[train_idx]

    X_test_fold = X.iloc[test_idx]
    y_test_fold = y.iloc[test_idx]

    est = fit_estimator(
        estimator=selected_spec["estimator"],
        X_train=X_train,
        y_train=y_train,
        use_sample_weight=selected_spec["use_sample_weight"],
    )

    perm = permutation_importance(
        est,
        X_test_fold,
        y_test_fold,
        scoring="average_precision",
        n_repeats=3,
        random_state=SEED + fold_id,
        n_jobs=-1,
    )

    for feature, imp_mean, imp_std in zip(feature_cols, perm.importances_mean, perm.importances_std):
        perm_tscv_rows.append({
            "fold": int(fold_id),
            "feature": feature,
            "importance_mean": float(imp_mean),
            "importance_std": float(imp_std),
            "source": "TimeSeriesSplit",
            "model": selected_model_name,
            "scoring": "average_precision",
        })

perm_tscv_raw_df = pd.DataFrame(perm_tscv_rows)

perm_tscv_df = (
    perm_tscv_raw_df
    .groupby(["feature", "model", "scoring"], as_index=False)
    .agg(
        importance_mean_tscv=("importance_mean", "mean"),
        importance_std_tscv=("importance_mean", "std"),
        folds=("fold", "nunique"),
    )
    .sort_values("importance_mean_tscv", ascending=False)
    .reset_index(drop=True)
)

print("\n=== Importância por permutação — média TSCV ===")
display(perm_tscv_df.head(20))

perm_tscv_df.to_csv(PERM_TSCV_FILE, index=False)
log(f"Importância por permutação TSCV salva: {PERM_TSCV_FILE}")


# -----------------------------
# 13) SUMMARY JSON
# -----------------------------

summary = {
    "notebook": "NB08",
    "purpose": "supervised_modeling_and_temporal_validation",
    "input_supervised_file": str(INPUT_SUPERVISED_FILE),
    "nb06_summary_file": str(NB06_SUMMARY_FILE),
    "dataset_rows": int(len(df)),
    "dataset_cols": int(df.shape[1]),
    "target_col": TARGET_COL,
    "target_distribution": target_distribution,
    "positive_rate": positive_rate,
    "feature_selection": feature_selection_info,
    "validation_primary": "TimeSeriesSplit",
    "n_splits_tscv": int(N_SPLITS_TSCV),
    "validation_secondary": "fixed_temporal_split",
    "fixed_split_fraction": float(FIXED_SPLIT_FRACTION),
    "split_sensitivity_fractions": SPLIT_SENSITIVITY_FRACTIONS,
    "models": {
        name: spec["notes"]
        for name, spec in MODEL_SPECS.items()
    },
    "selected_model": selected_model_name,
    "selected_model_criterion": (
        "Highest mean F1 in TimeSeriesSplit, with ROC-AUC and Average Precision as tie-breakers."
    ),
    "selected_tscv_summary": selected_tscv_summary,
    "tscv_summary": tscv_summary_df.to_dict("records"),
    "fixed_split_metrics": fixed_metrics_df.to_dict("records"),
    "split_sensitivity_metrics": split_sensitivity_df.to_dict("records"),
    "outputs": {
        "summary_file": str(SUMMARY_FILE),
        "tscv_fold_metrics_file": str(TSCV_FOLD_METRICS_FILE),
        "tscv_summary_file": str(TSCV_SUMMARY_FILE),
        "fixed_metrics_file": str(FIXED_METRICS_FILE),
        "split_sensitivity_file": str(SPLIT_SENSITIVITY_FILE),
        "permutation_importance_fixed_file": str(PERM_FIXED_FILE),
        "permutation_importance_tscv_file": str(PERM_TSCV_FILE),
        "selected_model_file": str(SELECTED_MODEL_FILE),
        "selected_feature_columns_file": str(SELECTED_FEATURES_FILE),
        "selected_test_predictions_file": str(SELECTED_TEST_PREDICTIONS_FILE),
    },
}

SUMMARY_FILE.write_text(
    json.dumps(summary, indent=2, ensure_ascii=False),
    encoding="utf-8",
)

log(f"Resumo final salvo: {SUMMARY_FILE}")


# -----------------------------
# 14) RESUMO FINAL NA TELA
# -----------------------------

print("\n=== RESUMO FINAL — NB08 ===")
print(f"Base supervisionada: {df.shape}")
print(f"Atributos selecionados: {len(feature_cols)}")
print(f"Distribuição do alvo: {target_distribution}")
print(f"Taxa positiva: {positive_rate:.4f}")
print()
print(f"Validação principal: TimeSeriesSplit com {N_SPLITS_TSCV} folds")
print(f"Modelo selecionado: {selected_model_name}")
print()
print("Resumo TSCV:")
display(tscv_summary_df)

print("\nCorte fixo 80/20:")
display(fixed_metrics_df)

print("\nSensibilidade por corte:")
display(split_sensitivity_df)

print("\nTop 15 — importância por permutação no corte fixo:")
display(perm_fixed_df.head(15))

print("\nTop 15 — importância por permutação média TSCV:")
display(perm_tscv_df.head(15))

Mounted at /content/drive
[Bootstrap] Atualizando repositório (git pull)...
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH = /content/drive/MyDrive/Mestrado/04-reports
MODELS_PATH = /content/drive/MyDrive/Mestrado/03-models
[NB08_supervised_modeling] Base supervisionada carregada: shape=(7267, 47)
[NB08_supervised_modeling] Arquivo: /content/drive/MyDrive/Mestrado/02-datasets/03-features/anticipation_supervised_dataset.parquet
[NB08_supervised_modeling] Resumo do NB06 carregado: /content/drive/MyDrive/Mestrado/04-reports/06_labeling_states_summary.json

=== Distribuição do alvo ===


,count
transition_target,
0,5280
1,1987


Taxa positiva: 0.2734

=== Tabela state × target ===


transition_target,0,1
state,,
BEFORE,979,1987
NORMAL,4301,0



=== Atributos selecionados ===
Total de atributos: 25
- n_events
- n_failed
- n_machines
- n_collections
- mean_priority
- mean_req_cpus
- mean_req_mem
- req_cpus_presence_rate
- req_mem_presence_rate
- event_FAIL_count
- event_SCHEDULE_count
- event_FINISH_count
- event_ENABLE_count
- event_LOST_count
- event_EVICT_count
- event_KILL_count
- fail_rate
- is_empty_window
- lag_1
- lag_2
- lag_3
- rolling_mean_1h
- rolling_std_1h
- pct_change
- zscore_expanding
[NB08_supervised_modeling] Lista de atributos salva: /content/drive/MyDrive/Mestrado/04-reports/08_selected_feature_columns.json

=== Modelos avaliados ===
LogisticRegression: StandardScaler + class_weight='balanced'
RandomForest: class_weight='balanced_subsample'
ExtraTrees: class_weight='balanced'
HistGradientBoosting: sample_weight=compute_sample_weight('balanced', y_train)
[NB08_supervised_modeling] TSCV — avaliando LogisticRegression
[NB08_supervised_modeling] TSCV — avaliando RandomForest
[NB08_supervised_modeling] TSCV — a

,validation,model,fold,train_start_pos,train_end_pos,test_start_pos,test_end_pos,train_rows,test_rows,train_positives,...,roc_auc,average_precision,brier_score,tn,fp,fn,tp,support,positives,positive_rate
0,TimeSeriesSplit,LogisticRegression,1,0,1211,1212,2422,1212,1211,36,...,0.537878,0.170177,0.397600,358,663,57,133,1211,190,0.156895
1,TimeSeriesSplit,LogisticRegression,2,0,2422,2423,3633,2423,1211,226,...,0.488867,0.057215,0.351628,474,665,31,41,1211,72,0.059455
2,TimeSeriesSplit,LogisticRegression,3,0,3633,3634,4844,3634,1211,298,...,0.727209,0.555866,0.252517,391,471,68,281,1211,349,0.288192
3,TimeSeriesSplit,LogisticRegression,4,0,4844,4845,6055,4845,1211,647,...,0.671128,0.621128,0.262857,200,398,45,568,1211,613,0.506193
4,TimeSeriesSplit,LogisticRegression,5,0,6055,6056,7266,6056,1211,1260,...,0.773463,0.823262,0.204327,177,307,59,668,1211,727,0.600330
5,TimeSeriesSplit,RandomForest,1,0,1211,1212,2422,1212,1211,36,...,0.544933,0.202939,0.139275,1021,0,190,0,1211,190,0.156895
6,TimeSeriesSplit,RandomForest,2,0,2422,2423,3633,2423,1211,226,...,0.501884,0.065405,0.069875,1137,2,72,0,1211,72,0.059455
7,TimeSeriesSplit,RandomForest,3,0,3633,3634,4844,3634,1211,298,...,0.711798,0.442369,0.211050,861,1,349,0,1211,349,0.288192
8,TimeSeriesSplit,RandomForest,4,0,4844,4845,6055,4845,1211,647,...,0.712751,0.702257,0.227275,420,178,298,315,1211,613,0.506193
9,TimeSeriesSplit,RandomForest,5,0,6055,6056,7266,6056,1211,1260,...,0.738774,0.779735,0.208451,303,181,201,526,1211,727,0.600330



=== Resumo TSCV — média e desvio-padrão ===


,model,accuracy_mean,accuracy_std,precision_mean,precision_std,recall_mean,recall_std,f1_mean,f1_std,roc_auc_mean,roc_auc_std,average_precision_mean,average_precision_std,brier_score_mean,brier_score_std,folds,test_positive_rate_mean,test_positive_rate_std
0,LogisticRegression,0.543518,0.127662,0.374390,0.267028,0.784007,0.151701,0.478004,0.289724,0.639709,0.122123,0.445530,0.321052,0.293786,0.078717,5,0.322213,0.228433
1,HistGradientBoosting,0.753922,0.109788,0.407223,0.272440,0.320127,0.395763,0.315336,0.363041,0.608766,0.093087,0.403360,0.294228,0.197583,0.076359,5,0.322213,0.228433
2,ExtraTrees,0.765813,0.121291,0.449139,0.364386,0.285318,0.383883,0.287160,0.377038,0.648020,0.116936,0.447132,0.312358,0.165005,0.053883,5,0.322213,0.228433
3,RandomForest,0.756895,0.132650,0.276587,0.380549,0.247478,0.346885,0.260647,0.361584,0.642028,0.109883,0.438541,0.308510,0.171185,0.065966,5,0.322213,0.228433


[NB08_supervised_modeling] Métricas TSCV por fold salvas: /content/drive/MyDrive/Mestrado/04-reports/08_tscv_metrics_by_fold.csv
[NB08_supervised_modeling] Resumo TSCV salvo: /content/drive/MyDrive/Mestrado/04-reports/08_tscv_summary.csv
[NB08_supervised_modeling] Modelo selecionado pelo critério TSCV: LogisticRegression
[NB08_supervised_modeling] Corte fixo 80/20 — avaliando LogisticRegression
[NB08_supervised_modeling] Corte fixo 80/20 — avaliando RandomForest
[NB08_supervised_modeling] Corte fixo 80/20 — avaliando ExtraTrees
[NB08_supervised_modeling] Corte fixo 80/20 — avaliando HistGradientBoosting

=== Corte temporal fixo 80/20 ===


,validation,split_fraction,model,train_rows,test_rows,train_positives,test_positives,train_positive_rate,test_positive_rate,accuracy,...,roc_auc,average_precision,brier_score,tn,fp,fn,tp,support,positives,positive_rate
0,fixed_temporal_split,0.8,LogisticRegression,5813,1454,1049,938,0.180458,0.645117,0.722146,...,0.765237,0.832725,0.193989,169,347,57,881,1454,938,0.645117
1,fixed_temporal_split,0.8,HistGradientBoosting,5813,1454,1049,938,0.180458,0.645117,0.702889,...,0.726542,0.809017,0.216297,259,257,175,763,1454,938,0.645117
2,fixed_temporal_split,0.8,ExtraTrees,5813,1454,1049,938,0.180458,0.645117,0.705640,...,0.746626,0.808899,0.194883,312,204,224,714,1454,938,0.645117
3,fixed_temporal_split,0.8,RandomForest,5813,1454,1049,938,0.180458,0.645117,0.669876,...,0.737856,0.813413,0.212753,345,171,309,629,1454,938,0.645117


[NB08_supervised_modeling] Métricas do corte fixo salvas: /content/drive/MyDrive/Mestrado/04-reports/08_fixed_split_metrics.csv
[NB08_supervised_modeling] Sensibilidade corte 0.60 — LogisticRegression
[NB08_supervised_modeling] Sensibilidade corte 0.60 — RandomForest
[NB08_supervised_modeling] Sensibilidade corte 0.60 — ExtraTrees
[NB08_supervised_modeling] Sensibilidade corte 0.60 — HistGradientBoosting
[NB08_supervised_modeling] Sensibilidade corte 0.70 — LogisticRegression
[NB08_supervised_modeling] Sensibilidade corte 0.70 — RandomForest
[NB08_supervised_modeling] Sensibilidade corte 0.70 — ExtraTrees
[NB08_supervised_modeling] Sensibilidade corte 0.70 — HistGradientBoosting
[NB08_supervised_modeling] Sensibilidade corte 0.80 — LogisticRegression
[NB08_supervised_modeling] Sensibilidade corte 0.80 — RandomForest
[NB08_supervised_modeling] Sensibilidade corte 0.80 — ExtraTrees
[NB08_supervised_modeling] Sensibilidade corte 0.80 — HistGradientBoosting

=== Sensibilidade por corte tem

,validation,split_fraction,model,train_rows,test_rows,train_positives,test_positives,train_positive_rate,test_positive_rate,accuracy,...,roc_auc,average_precision,brier_score,tn,fp,fn,tp,support,positives,positive_rate
0,split_sensitivity,0.6,LogisticRegression,4360,2907,447,1540,0.102523,0.529756,0.648779,...,0.720449,0.717294,0.227536,523,844,177,1363,2907,1540,0.529756
1,split_sensitivity,0.6,RandomForest,4360,2907,447,1540,0.102523,0.529756,0.573443,...,0.739263,0.727492,0.255652,1250,117,1123,417,2907,1540,0.529756
2,split_sensitivity,0.6,ExtraTrees,4360,2907,447,1540,0.102523,0.529756,0.602683,...,0.741903,0.718505,0.230398,1170,197,958,582,2907,1540,0.529756
3,split_sensitivity,0.6,HistGradientBoosting,4360,2907,447,1540,0.102523,0.529756,0.656003,...,0.734526,0.731455,0.247778,1036,331,669,871,2907,1540,0.529756
4,split_sensitivity,0.7,LogisticRegression,5086,2181,689,1298,0.135470,0.595140,0.659789,...,0.691725,0.743722,0.233433,243,640,102,1196,2181,1298,0.595140
5,split_sensitivity,0.7,RandomForest,5086,2181,689,1298,0.135470,0.595140,0.624943,...,0.711385,0.770209,0.230375,611,272,546,752,2181,1298,0.595140
6,split_sensitivity,0.7,ExtraTrees,5086,2181,689,1298,0.135470,0.595140,0.649702,...,0.705534,0.762460,0.218774,537,346,418,880,2181,1298,0.595140
7,split_sensitivity,0.7,HistGradientBoosting,5086,2181,689,1298,0.135470,0.595140,0.658414,...,0.692126,0.759497,0.250492,458,425,320,978,2181,1298,0.595140
8,split_sensitivity,0.8,LogisticRegression,5813,1454,1049,938,0.180458,0.645117,0.722146,...,0.765237,0.832725,0.193989,169,347,57,881,1454,938,0.645117
9,split_sensitivity,0.8,RandomForest,5813,1454,1049,938,0.180458,0.645117,0.669876,...,0.737856,0.813413,0.212753,345,171,309,629,1454,938,0.645117


[NB08_supervised_modeling] Sensibilidade de cortes salva: /content/drive/MyDrive/Mestrado/04-reports/08_split_sensitivity_metrics.csv
[NB08_supervised_modeling] Predições do modelo selecionado salvas: /content/drive/MyDrive/Mestrado/04-reports/08_selected_model_test_predictions.parquet
[NB08_supervised_modeling] Modelo selecionado salvo: /content/drive/MyDrive/Mestrado/03-models/08_selected_model.joblib
[NB08_supervised_modeling] Calculando importância por permutação no corte fixo para LogisticRegression

=== Importância por permutação — corte fixo ===


,feature,importance_mean,importance_std,source,model,scoring
0,rolling_mean_1h,0.075226,0.011436,fixed_temporal_split,LogisticRegression,average_precision
1,rolling_std_1h,0.014841,0.003207,fixed_temporal_split,LogisticRegression,average_precision
2,n_machines,0.010356,0.003094,fixed_temporal_split,LogisticRegression,average_precision
3,fail_rate,0.008198,0.003407,fixed_temporal_split,LogisticRegression,average_precision
4,n_events,0.005346,0.000456,fixed_temporal_split,LogisticRegression,average_precision
5,event_EVICT_count,0.004809,0.000393,fixed_temporal_split,LogisticRegression,average_precision
6,event_SCHEDULE_count,0.003699,0.000495,fixed_temporal_split,LogisticRegression,average_precision
7,mean_priority,0.003307,0.001412,fixed_temporal_split,LogisticRegression,average_precision
8,lag_1,0.002290,0.001768,fixed_temporal_split,LogisticRegression,average_precision
9,lag_2,0.001008,0.001544,fixed_temporal_split,LogisticRegression,average_precision


[NB08_supervised_modeling] Importância por permutação no corte fixo salva: /content/drive/MyDrive/Mestrado/04-reports/08_permutation_importance_fixed.csv
[NB08_supervised_modeling] Calculando importância por permutação nos folds TSCV para LogisticRegression

=== Importância por permutação — média TSCV ===


,feature,model,scoring,importance_mean_tscv,importance_std_tscv,folds
0,zscore_expanding,LogisticRegression,average_precision,0.072077,0.117047,5
1,rolling_std_1h,LogisticRegression,average_precision,0.035508,0.036427,5
2,rolling_mean_1h,LogisticRegression,average_precision,0.031813,0.057214,5
3,fail_rate,LogisticRegression,average_precision,0.012401,0.074186,5
4,lag_1,LogisticRegression,average_precision,0.012217,0.015221,5
5,lag_2,LogisticRegression,average_precision,0.012151,0.020544,5
6,lag_3,LogisticRegression,average_precision,0.008707,0.017248,5
7,n_collections,LogisticRegression,average_precision,0.006810,0.013838,5
8,n_machines,LogisticRegression,average_precision,0.003947,0.009512,5
9,event_FINISH_count,LogisticRegression,average_precision,0.001973,0.006949,5


[NB08_supervised_modeling] Importância por permutação TSCV salva: /content/drive/MyDrive/Mestrado/04-reports/08_permutation_importance_tscv.csv
[NB08_supervised_modeling] Resumo final salvo: /content/drive/MyDrive/Mestrado/04-reports/08_supervised_modeling_summary.json

=== RESUMO FINAL — NB08 ===
Base supervisionada: (7267, 47)
Atributos selecionados: 25
Distribuição do alvo: {'0': 5280, '1': 1987}
Taxa positiva: 0.2734

Validação principal: TimeSeriesSplit com 5 folds
Modelo selecionado: LogisticRegression

Resumo TSCV:


,model,accuracy_mean,accuracy_std,precision_mean,precision_std,recall_mean,recall_std,f1_mean,f1_std,roc_auc_mean,roc_auc_std,average_precision_mean,average_precision_std,brier_score_mean,brier_score_std,folds,test_positive_rate_mean,test_positive_rate_std
0,LogisticRegression,0.543518,0.127662,0.374390,0.267028,0.784007,0.151701,0.478004,0.289724,0.639709,0.122123,0.445530,0.321052,0.293786,0.078717,5,0.322213,0.228433
1,HistGradientBoosting,0.753922,0.109788,0.407223,0.272440,0.320127,0.395763,0.315336,0.363041,0.608766,0.093087,0.403360,0.294228,0.197583,0.076359,5,0.322213,0.228433
2,ExtraTrees,0.765813,0.121291,0.449139,0.364386,0.285318,0.383883,0.287160,0.377038,0.648020,0.116936,0.447132,0.312358,0.165005,0.053883,5,0.322213,0.228433
3,RandomForest,0.756895,0.132650,0.276587,0.380549,0.247478,0.346885,0.260647,0.361584,0.642028,0.109883,0.438541,0.308510,0.171185,0.065966,5,0.322213,0.228433



Corte fixo 80/20:


,validation,split_fraction,model,train_rows,test_rows,train_positives,test_positives,train_positive_rate,test_positive_rate,accuracy,...,roc_auc,average_precision,brier_score,tn,fp,fn,tp,support,positives,positive_rate
0,fixed_temporal_split,0.8,LogisticRegression,5813,1454,1049,938,0.180458,0.645117,0.722146,...,0.765237,0.832725,0.193989,169,347,57,881,1454,938,0.645117
1,fixed_temporal_split,0.8,HistGradientBoosting,5813,1454,1049,938,0.180458,0.645117,0.702889,...,0.726542,0.809017,0.216297,259,257,175,763,1454,938,0.645117
2,fixed_temporal_split,0.8,ExtraTrees,5813,1454,1049,938,0.180458,0.645117,0.705640,...,0.746626,0.808899,0.194883,312,204,224,714,1454,938,0.645117
3,fixed_temporal_split,0.8,RandomForest,5813,1454,1049,938,0.180458,0.645117,0.669876,...,0.737856,0.813413,0.212753,345,171,309,629,1454,938,0.645117



Sensibilidade por corte:


,validation,split_fraction,model,train_rows,test_rows,train_positives,test_positives,train_positive_rate,test_positive_rate,accuracy,...,roc_auc,average_precision,brier_score,tn,fp,fn,tp,support,positives,positive_rate
0,split_sensitivity,0.6,LogisticRegression,4360,2907,447,1540,0.102523,0.529756,0.648779,...,0.720449,0.717294,0.227536,523,844,177,1363,2907,1540,0.529756
1,split_sensitivity,0.6,RandomForest,4360,2907,447,1540,0.102523,0.529756,0.573443,...,0.739263,0.727492,0.255652,1250,117,1123,417,2907,1540,0.529756
2,split_sensitivity,0.6,ExtraTrees,4360,2907,447,1540,0.102523,0.529756,0.602683,...,0.741903,0.718505,0.230398,1170,197,958,582,2907,1540,0.529756
3,split_sensitivity,0.6,HistGradientBoosting,4360,2907,447,1540,0.102523,0.529756,0.656003,...,0.734526,0.731455,0.247778,1036,331,669,871,2907,1540,0.529756
4,split_sensitivity,0.7,LogisticRegression,5086,2181,689,1298,0.135470,0.595140,0.659789,...,0.691725,0.743722,0.233433,243,640,102,1196,2181,1298,0.595140
5,split_sensitivity,0.7,RandomForest,5086,2181,689,1298,0.135470,0.595140,0.624943,...,0.711385,0.770209,0.230375,611,272,546,752,2181,1298,0.595140
6,split_sensitivity,0.7,ExtraTrees,5086,2181,689,1298,0.135470,0.595140,0.649702,...,0.705534,0.762460,0.218774,537,346,418,880,2181,1298,0.595140
7,split_sensitivity,0.7,HistGradientBoosting,5086,2181,689,1298,0.135470,0.595140,0.658414,...,0.692126,0.759497,0.250492,458,425,320,978,2181,1298,0.595140
8,split_sensitivity,0.8,LogisticRegression,5813,1454,1049,938,0.180458,0.645117,0.722146,...,0.765237,0.832725,0.193989,169,347,57,881,1454,938,0.645117
9,split_sensitivity,0.8,RandomForest,5813,1454,1049,938,0.180458,0.645117,0.669876,...,0.737856,0.813413,0.212753,345,171,309,629,1454,938,0.645117



Top 15 — importância por permutação no corte fixo:


,feature,importance_mean,importance_std,source,model,scoring
0,rolling_mean_1h,0.075226,0.011436,fixed_temporal_split,LogisticRegression,average_precision
1,rolling_std_1h,0.014841,0.003207,fixed_temporal_split,LogisticRegression,average_precision
2,n_machines,0.010356,0.003094,fixed_temporal_split,LogisticRegression,average_precision
3,fail_rate,0.008198,0.003407,fixed_temporal_split,LogisticRegression,average_precision
4,n_events,0.005346,0.000456,fixed_temporal_split,LogisticRegression,average_precision
5,event_EVICT_count,0.004809,0.000393,fixed_temporal_split,LogisticRegression,average_precision
6,event_SCHEDULE_count,0.003699,0.000495,fixed_temporal_split,LogisticRegression,average_precision
7,mean_priority,0.003307,0.001412,fixed_temporal_split,LogisticRegression,average_precision
8,lag_1,0.002290,0.001768,fixed_temporal_split,LogisticRegression,average_precision
9,lag_2,0.001008,0.001544,fixed_temporal_split,LogisticRegression,average_precision



Top 15 — importância por permutação média TSCV:


,feature,model,scoring,importance_mean_tscv,importance_std_tscv,folds
0,zscore_expanding,LogisticRegression,average_precision,0.072077,0.117047,5
1,rolling_std_1h,LogisticRegression,average_precision,0.035508,0.036427,5
2,rolling_mean_1h,LogisticRegression,average_precision,0.031813,0.057214,5
3,fail_rate,LogisticRegression,average_precision,0.012401,0.074186,5
4,lag_1,LogisticRegression,average_precision,0.012217,0.015221,5
5,lag_2,LogisticRegression,average_precision,0.012151,0.020544,5
6,lag_3,LogisticRegression,average_precision,0.008707,0.017248,5
7,n_collections,LogisticRegression,average_precision,0.006810,0.013838,5
8,n_machines,LogisticRegression,average_precision,0.003947,0.009512,5
9,event_FINISH_count,LogisticRegression,average_precision,0.001973,0.006949,5


## 11. Conclusão da Etapa

A etapa de modelagem supervisionada e validação temporal foi concluída com sucesso, utilizando a base revisada de antecipação produzida no NB06.

A base utilizada nesta execução corresponde ao cenário `primary_K24_H12`, no qual:

- \(K = 24\) janelas define a região pré-crítica;
- \(H = 12\) janelas define o horizonte supervisionado de antecipação;
- o horizonte preditivo corresponde a 60 minutos;
- a janela pré-crítica corresponde a 120 minutos.

A base supervisionada final apresentou:

- 7.267 amostras;
- 47 colunas de entrada;
- 5.280 exemplos negativos;
- 1.987 exemplos positivos;
- taxa positiva de 27,34%.

Foram selecionados 25 atributos explicativos, excluindo identificadores, campos de tempo absoluto, estados operacionais, variáveis futuras, campos auxiliares da rotulagem e o rótulo instantâneo `is_critical`.

Essa seleção preserva a coerência causal da modelagem, pois os atributos utilizados representam informações disponíveis até a janela corrente.

---

## 12. Validação Temporal Principal

A avaliação principal foi realizada por validação temporal por partições progressivas (`TimeSeriesSplit`) com 5 folds.

Esse protocolo foi adotado como principal por ser mais compatível com o uso operacional em séries temporais, já que respeita a ordem temporal dos dados e avalia o modelo em blocos futuros sucessivos.

A Regressão Logística foi selecionada como modelo principal pelo critério de maior F1 médio em TSCV, com ROC-AUC e Average Precision como critérios de desempate.

No TSCV, a Regressão Logística apresentou:

- accuracy média = 0,544;
- precision média = 0,374;
- recall médio = 0,784;
- F1 médio = 0,478;
- ROC-AUC médio = 0,640;
- Average Precision médio = 0,446;
- Brier Score médio = 0,294.

Os desvios-padrão foram elevados, especialmente em F1, precision e Average Precision. Isso indica instabilidade temporal relevante entre folds e confirma que o desempenho deve ser interpretado com cautela.

A avaliação temporal principal, portanto, não sustenta uma conclusão de desempenho robusto, mas indica a presença de sinal preditivo moderado, especialmente associado ao recall.

---

## 13. Comparação entre Modelos

Os modelos avaliados foram:

- Regressão Logística;
- Random Forest;
- ExtraTrees;
- HistGradientBoosting.

A Regressão Logística utilizou `StandardScaler` e `class_weight='balanced'`.

Os modelos Random Forest e ExtraTrees foram avaliados sem padronização, com tratamento de desbalanceamento por `class_weight`.

O HistGradientBoosting foi avaliado com `sample_weight`, calculado a partir do balanceamento das classes no conjunto de treinamento.

No TSCV, a Regressão Logística obteve o maior F1 médio. Os modelos baseados em árvores apresentaram maior accuracy média em alguns casos, mas menor recall e menor F1 médio.

Esse resultado sugere que, para a formulação revisada, a Regressão Logística manteve comportamento mais adequado ao objetivo de antecipação, por priorizar recall e produzir maior F1 médio sob validação temporal.

---

## 14. Corte Temporal Fixo 80/20

O corte temporal fixo 80/20 foi mantido apenas como análise complementar.

Nesse corte, a distribuição de classes apresentou forte mudança entre treinamento e teste:

- treino: 5.813 amostras, com 1.049 positivos;
- taxa positiva no treino: 18,05%;
- teste: 1.454 amostras, com 938 positivos;
- taxa positiva no teste: 64,51%.

Essa diferença confirma que o corte fixo não deve ser tratado como estimativa principal de generalização operacional.

Ainda assim, no corte fixo, a Regressão Logística apresentou:

- accuracy = 0,722;
- precision = 0,717;
- recall = 0,939;
- F1 = 0,813;
- ROC-AUC = 0,765;
- Average Precision = 0,833;
- Brier Score = 0,194.

Esse desempenho elevado indica que o modelo consegue capturar padrões relevantes nesse recorte, mas a interpretação deve considerar a forte concentração de exemplos positivos no conjunto de teste.

---

## 15. Sensibilidade aos Cortes Temporais

Foram avaliados cortes cronológicos adicionais:

- 60/40;
- 70/30;
- 80/20.

A Regressão Logística manteve recall elevado nos três cortes:

- corte 60/40: recall = 0,885 e F1 = 0,728;
- corte 70/30: recall = 0,921 e F1 = 0,763;
- corte 80/20: recall = 0,939 e F1 = 0,813.

Entretanto, os cortes também mostraram mudança progressiva na proporção de positivos no conjunto de teste:

- 60/40: teste com 52,98% de positivos;
- 70/30: teste com 59,51% de positivos;
- 80/20: teste com 64,51% de positivos.

Isso indica que os resultados do corte fixo são sensíveis à posição temporal do split e devem ser interpretados como análise complementar, não como evidência principal.

---

## 16. Importância dos Atributos

A importância por permutação foi calculada em dois contextos:

1. no corte temporal fixo;
2. nos folds do `TimeSeriesSplit`.

No corte fixo, os atributos mais relevantes para a Regressão Logística foram:

- `rolling_mean_1h`;
- `rolling_std_1h`;
- `n_machines`;
- `fail_rate`;
- `n_events`.

Na média dos folds do TSCV, os atributos mais relevantes foram:

- `zscore_expanding`;
- `rolling_std_1h`;
- `rolling_mean_1h`;
- `fail_rate`;
- `lag_1`.

Esses resultados reforçam que a dinâmica recente da taxa de falha, sua variabilidade local e medidas acumuladas de desvio temporal concentram parte importante do sinal preditivo.

No entanto, como a importância variou entre o corte fixo e o TSCV, essa análise deve ser interpretada como evidência exploratória de interpretabilidade, e não como prova definitiva de estabilidade dos atributos.

---

## 17. Artefatos Gerados

Nesta execução, foram gerados os seguintes artefatos:

- `08_supervised_modeling_summary.json`;
- `08_tscv_metrics_by_fold.csv`;
- `08_tscv_summary.csv`;
- `08_fixed_split_metrics.csv`;
- `08_split_sensitivity_metrics.csv`;
- `08_permutation_importance_fixed.csv`;
- `08_permutation_importance_tscv.csv`;
- `08_selected_feature_columns.json`;
- `08_selected_model.joblib`;
- `08_selected_model_test_predictions.parquet`.

O arquivo `08_selected_model_test_predictions.parquet` contém as probabilidades previstas do modelo selecionado no corte fixo e será utilizado no NB09 para geração das figuras finais de probabilidades ao longo do tempo.

---

## 18. Encaminhamento do Pipeline

A partir desta execução, o artigo e a dissertação devem tratar a validação temporal por `TimeSeriesSplit` como resultado principal.

O corte temporal fixo deve ser apresentado como análise complementar, útil para visualização e inspeção das probabilidades previstas, mas não como evidência principal de generalização.

O próximo passo é revisar o NB09 para:

- usar as predições geradas pelo NB08 revisado;
- atualizar as figuras finais;
- incluir avaliação de calibração probabilística;
- gerar Brier Score e curva de calibração;
- ajustar a interpretação das probabilidades como escores de risco, e não como probabilidades plenamente calibradas.